src/Storages/ObjectStorage/DataLakes/Iceberg/IcebergWrites.cpp

```diff
@@ -0,0 +1,772 @@
+            {
+                throw Exception(ErrorCodes::BAD_ARGUMENTS, "Unknown transform {}", transform_name);
+            }
+
+            auto argument_width = transform_name.length() - 2 - argument_start;
+            std::string argument_string_representation = transform_name.substr(argument_start + 1, argument_width);
+            size_t argument;
+            bool parsed = DB::tryParse<size_t>(argument, argument_string_representation);
+
+            if (!parsed)
+            {
+                throw Exception(ErrorCodes::BAD_ARGUMENTS, "Unknown transform {}", transform_name);
+            }
+
+            transform_param = argument;
+            if (transform_name.starts_with("truncate"))
+            {
+                transform = factory.get("icebergTruncate", context);
+            }
+            else if (transform_name.starts_with("bucket"))
+            {
+                transform = factory.get("icebergBucket", context);
+            }
+        }
+
+        functions.push_back(transform);
+        function_params.push_back(transform_param);
+        columns_to_apply.push_back(column_name);
+    }
+}
+
+size_t ChunkPartitioner::PartitionKeyHasher::operator()(const PartitionKey & key) const
+{
+    size_t result = 0;
+    for (const auto & part_key : key)
+        result ^= hasher(part_key.dump());
+    return result;
+}
+
+std::unordered_map<ChunkPartitioner::PartitionKey, Chunk, ChunkPartitioner::PartitionKeyHasher>
+ChunkPartitioner::participateChunk(const Chunk & chunk)
+{
+    std::unordered_map<String, ColumnWithTypeAndName> name_to_column;
+    for (size_t i = 0; i < sample_block.columns(); ++i)
+    {
+        auto column_ptr = chunk.getColumns()[i];
+        auto column_name = sample_block.getNames()[i];
+        name_to_column[column_name] = ColumnWithTypeAndName(column_ptr, sample_block.getDataTypes()[i], column_name);
+    }
+
+    std::vector<ChunkPartitioner::PartitionKey> transform_results(chunk.getNumRows());
+    for (size_t transform_ind = 0; transform_ind < functions.size(); ++transform_ind)
+    {
+        ColumnsWithTypeAndName arguments;
+        if (function_params[transform_ind].has_value())
+        {
+            auto type = std::make_shared<DataTypeUInt64>();
+            auto column_value = ColumnUInt64::create();
+            column_value->insert(*function_params[transform_ind]);
+            auto const_column = ColumnConst::create(std::move(column_value), chunk.getNumRows());
+            arguments.push_back(ColumnWithTypeAndName(const_column->clone(), type, "#"));
+        }
+        arguments.push_back(name_to_column[columns_to_apply[transform_ind]]);
+        auto result
+            = functions[transform_ind]->build(arguments)->execute(arguments, std::make_shared<DataTypeString>(), chunk.getNumRows(), false);
+        for (size_t i = 0; i < chunk.getNumRows(); ++i)
+        {
+            Field field;
+            result->get(i, field);
+            transform_results[i].push_back(field.dump());
```

Comment:
Just build selector and use column->scatter, take a lot at MergeTreeDataWriter code.